pandas/core/groupby.py

```diff
@@ -3740,9 +3740,31 @@ def _reindex_output(self, result):
             return result
 
         levels_list = [ping.group_index for ping in groupings]
-        index = MultiIndex.from_product(levels_list, names=self.grouper.names)
-        d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
-        return result.reindex(**d).sortlevel(axis=self.axis)
+        index, _ = MultiIndex.from_product(
+            levels_list, names=self.grouper.names).sortlevel()
+
+        if self.as_index:
+            d = {self.obj._get_axis_name(self.axis): index, 'copy': False}
+            return result.reindex(**d)
+
+        # GH 13204
```

Comment:
doesn't something like this solve the issue? This is conceptually what ``as_index=False`` is anyhow.
```
3760)_reindex_output()
-> indexer = self.grouper.result_index.get_indexer(index)
(Pdb) result.set_index(['cat','A']).reindex(index).reset_index()
   cat   A      B
0    1  10  101.0
1    1  11    NaN
2    2  10    NaN
3    2  11  205.0
4    3  10    NaN
5    3  11    NaN
```